lib/src/main/java/com/squareup/cycler/Update.kt

```diff
@@ -120,49 +166,68 @@ class Update<I : Any>(private val oldRecyclerData: RecyclerData<I>) {
   internal fun generateUpdateWork(itemComparator: ItemComparator<I>?): UpdateWork {
 
     val extraItemChanged = oldRecyclerData.extraItem != extraItem
-    val refreshAllNeeded = dataReplaced && itemComparator == null
 
     val asyncWork = mutableListOf<() -> Unit>()
     val notifications = mutableListOf<(Adapter<*>) -> Unit>()
 
-    if (refreshAllNeeded) {
-      @SuppressLint("NotifyDataSetChanged")
-      notifications += { adapter -> adapter.notifyDataSetChanged() }
-    } else {
+    if (extraItemChanged) {
+      // We notify first the extraItem change, so we use its position in oldRecyclerData.
+      notifications += notifyChangesExtraItem()
+    }
 
-      if (extraItemChanged) {
-        // We notify first the extraItem change, so we use its position in oldRecyclerData.
-        notifications += notifyChangesExtraItem()
+    when {
+      dataPopulated -> {
+        val work = generateUpdateLambdas(populateMode, itemComparator)
+        work.first?.let { asyncWork.add(it) }
+        notifications.add(work.second)
       }
-
-      when {
-        dataReplaced -> {
-          // refreshAllNeeded == false => itemComparator != null. We are going async.
-          lateinit var diffResult: DiffResult
-          asyncWork += {
-            diffResult = calculateDataChanges(itemComparator!!)
-          }
-          notifications += { adapter -> diffResult.dispatchUpdatesTo(adapter) }
-        }
-        dataAdded -> {
-          notifications += { adapter ->
-            val positionAt = oldRecyclerData.data.size
-            val count = addedChunks.asSequence()
-              .map(List<I>::size)
-              .sum()
-            adapter.notifyItemRangeInserted(positionAt, count)
-          }
+      dataReplaced -> {
+        val work = generateUpdateLambdas(replaceMode, itemComparator)
+        work.first?.let { asyncWork.add(it) }
+        notifications.add(work.second)
+      }
+      dataAdded -> {
+        notifications += { adapter ->
+          val positionAt = oldRecyclerData.data.size
+          val count = addedChunks.asSequence()
+            .map(List<I>::size)
+            .sum()
+          adapter.notifyItemRangeInserted(positionAt, count)
         }
       }
     }
 
     return UpdateWork(
-        allowSynchronousUpdate = allowSynchronousUpdate,
+        populateMode = populateMode,
+        replaceMode = replaceMode,
+        populating = dataPopulated,
+        replacing = dataReplaced,
         asyncWork = asyncWork,
         notifications = notifications
     )
   }
 
+  @SuppressLint("NotifyDataSetChanged")
+  internal fun generateUpdateLambdas(
+    updateMode: UpdateMode,
+    itemComparator: ItemComparator<I>?
```

Comment:
Let's not call this method if comparator is null, if it's null (and we are replacing, just change do the populate thing).